src/firebase_functions/identity_fn.py

```diff
@@ -0,0 +1,433 @@
+    """
+    The user's password salt (base64-encoded).
+    """
+
+    custom_claims: dict[str, _typing.Any] | None
+    """
+    The user's custom claims object if available.
+    """
+
+    tenant_id: str | None
+    """
+    The ID of the tenant the user belongs to, if available.
+    """
+
+    tokens_valid_after_time: _dt.datetime | None
+    """The date the user's tokens are valid after."""
+
+    multi_factor: AuthMultiFactorSettings | None
+    """The multi-factor related properties for the current user, if available."""
+
+
+@_dataclasses.dataclass(frozen=True)
+class AdditionalUserInfo:
+    """
+    The additional user info component of the auth event context.
+    """
+
+    provider_id: str
+    """The provider identifier."""
+
+    profile: dict[str, _typing.Any] | None
+    """The user's profile data as a dictionary."""
+
+    username: str | None
+    """The user's username, if available."""
+
+    is_new_user: bool
+    """A boolean indicating if the user is new or not."""
+
+
+@_dataclasses.dataclass(frozen=True)
+class Credential:
+    """
+    The credential component of the auth event context.
+    """
+
+    claims: dict[str, _typing.Any] | None
+    """The user's claims data as a dictionary."""
+
+    id_token: str | None
+    """The user's ID token."""
+
+    access_token: str | None
+    """The user's access token."""
+
+    refresh_token: str | None
+    """The user's refresh token."""
+
+    expiration_time: _dt.datetime | None
+    """The expiration time of the user's access token."""
+
+    secret: str | None
+    """The user's secret."""
+
+    provider_id: str
+    """The provider identifier."""
+
+    sign_in_method: str
+    """The user's sign-in method."""
+
+
+@_dataclasses.dataclass(frozen=True)
+class AuthBlockingEvent:
+    """
+    Defines an auth event for identitytoolkit v2 auth blocking events.
+    """
+
+    data: AuthUserRecord
+    """
+    The UserRecord passed to auth blocking Cloud Functions from the identity platform.
+    """
+
+    locale: str | None
+    """
+    The application locale. You can set the locale using the client SDK, 
+    or by passing the locale header in the REST API.
+    Example: 'fr' or 'sv-SE'
+    """
+
+    event_id: str
+    """
+    The event's unique identifier.
+    Example: 'rWsyPtolplG2TBFoOkkgyg'
+    """
+
+    ip_address: str
+    """
+    The IP address of the device the end user is registering or signing in from.
+    Example: '114.14.200.1'
+    """
+
+    user_agent: str
+    """
+    The user agent triggering the blocking function.
+    Example: 'Mozilla/5.0 (X11; Linux x86_64)'
+    """
+
+    additional_user_info: AdditionalUserInfo
+    """An object containing information about the user."""
+
+    credential: Credential | None
+    """An object containing information about the user's credential."""
+
+    timestamp: _dt.datetime
+    """
+    The time the event was triggered."""
+
+
+class BeforeCreateResponse(_typing.TypedDict, total=False):
```

Comment:
q: What's the difference between typedict vs dataclass 🤔 

should we prefer one style over another here?

Reply:
answer: typedict allows us to distinguish undefined vs null (in js)